bugbug/models/accessibility.py

```diff
@@ -0,0 +1,145 @@
+
+import xgboost
+from dateutil.relativedelta import relativedelta
+from imblearn.over_sampling import BorderlineSMOTE
+from imblearn.pipeline import Pipeline as ImblearnPipeline
+from sklearn.compose import ColumnTransformer
+from sklearn.feature_extraction import DictVectorizer
+from sklearn.pipeline import Pipeline
+
+from bugbug import bug_features, bugzilla, feature_cleanup, utils
+from bugbug.model import BugModel
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+class AccessibilityModel(BugModel):
+    def __init__(self, lemmatization=False):
+        BugModel.__init__(self, lemmatization)
+
+        self.calculate_importance = False
+
+        feature_extractors = [
+            bug_features.HasSTR(),
+            bug_features.Severity(),
+            bug_features.Keywords({"access"}),
+            bug_features.HasAttachment(),
+            bug_features.Product(),
+            bug_features.FiledVia(),
+            bug_features.HasImageAttachment(),
+            bug_features.Component(),
+        ]
+
+        cleanup_functions = [
+            feature_cleanup.fileref(),
+            feature_cleanup.url(),
+            feature_cleanup.synonyms(),
+        ]
+
+        self.extraction_pipeline = Pipeline(
+            [
+                (
+                    "bug_extractor",
+                    bug_features.BugExtractor(
+                        feature_extractors,
+                        cleanup_functions,
+                        rollback=True,
+                    ),
+                ),
+            ]
+        )
+
+        self.clf = ImblearnPipeline(
+            [
+                (
+                    "union",
+                    ColumnTransformer(
+                        [
+                            ("data", DictVectorizer(), "data"),
+                            ("title", self.text_vectorizer(min_df=0.0001), "title"),
+                            (
+                                "comments",
+                                self.text_vectorizer(min_df=0.0001),
+                                "comments",
+                            ),
+                        ]
+                    ),
+                ),
+                ("sampler", BorderlineSMOTE(random_state=0)),
+                (
+                    "estimator",
+                    xgboost.XGBClassifier(
+                        n_jobs=utils.get_physical_cpu_count(), scale_pos_weight=10
+                    ),
+                ),
+            ]
+        )
+
+    @staticmethod
+    def __is_accessibility_bug(bug):
+        """Check if a bug is an accessibility bug."""
+        return bug["cf_accessibility_severity"] != "---" or "access" in bug["keywords"]
+
+    @staticmethod
+    def __get_access_bugs_ids():
+        """Get accessibility related bugs not older than 4 years and 6 months ago."""
+        four_years_and_six_months_ago = datetime.utcnow() - relativedelta(
+            years=4, months=6
+        )
+        params = {
```

Comment:
Can you please try to restrict it to the products that are supported?

Reply:
Training with the `Severity` feature, WDYT?
<details>
<summary>Classification Report</summary>

```
2023-12-21 22:39:14,562:ERROR:bugbug.bug_snapshot:1868902 is not in [1868902], for field duplicates, in bug 1864760
2023-12-21 22:40:55,731:INFO:bugbug.model:X: (52795, 4), y: (52795,)
2023-12-21 22:49:34,082:INFO:bugbug.model:Cross Validation scores:
2023-12-21 22:49:34,082:INFO:bugbug.model:Accuracy: f0.9876460065242554 (+/- 0.0014617493721296902)
2023-12-21 22:49:34,082:INFO:bugbug.model:Precision: f0.7396711053144303 (+/- 0.03755583626989041)
2023-12-21 22:49:34,082:INFO:bugbug.model:Recall: f0.6528542451463345 (+/- 0.06396957952549281)
2023-12-21 22:49:34,083:INFO:bugbug.model:X_train: (47515, 4), y_train: (47515,)
2023-12-21 22:49:34,083:INFO:bugbug.model:X_test: (5280, 4), y_test: (5280,)
2023-12-21 22:51:24,417:INFO:bugbug.model:Number of features: 37815
2023-12-21 22:51:24,417:INFO:bugbug.model:Model trained
2023-12-21 22:51:24,508:INFO:bugbug.model:Training Set scores:
                   pre       rec       spe        f1       geo       iba       sup

          1       0.93      1.00      1.00      0.97      1.00      1.00      1017
          0       1.00      1.00      1.00      1.00      1.00      1.00     46498

avg / total       1.00      1.00      1.00      1.00      1.00      1.00     47515

2023-12-21 22:51:47,173:INFO:bugbug.model:Test Set scores:
2023-12-21 22:51:49,775:INFO:bugbug.model:No confidence threshold - 5280 classified
                   pre       rec       spe        f1       geo       iba       sup

          1       0.80      0.70      1.00      0.75      0.84      0.68       122
          0       0.99      1.00      0.70      0.99      0.84      0.72      5158

avg / total       0.99      0.99      0.71      0.99      0.84      0.72      5280

╒════════════╤═════════════════╤═════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │
╞════════════╪═════════════════╪═════════════════╡
│ 1 (Actual) │              86 │              36 │
├────────────┼─────────────────┼─────────────────┤
│ 0 (Actual) │              22 │            5136 │
╘════════════╧═════════════════╧═════════════════╛

2023-12-21 22:51:52,550:INFO:bugbug.model:
Confidence threshold > 0.1 - 5280 classified
Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
Sensitivity is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
                          pre       rec       spe        f1       geo       iba       sup

                 1       0.48      0.83      0.98      0.61      0.90      0.80       122
                 0       1.00      0.98      0.83      0.99      0.90      0.82      5158
__NOT_CLASSIFIED__       0.00      0.00      1.00      0.00      0.00      0.00         0

       avg / total       0.98      0.98      0.83      0.98      0.90      0.82      5280

╒════════════╤═════════════════╤═════════════════╤══════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │   Not classified │
╞════════════╪═════════════════╪═════════════════╪══════════════════╡
│ 1 (Actual) │             101 │              21 │                0 │
├────────────┼─────────────────┼─────────────────┼──────────────────┤
│ 0 (Actual) │             110 │            5048 │                0 │
╘════════════╧═════════════════╧═════════════════╧══════════════════╛

2023-12-21 22:51:55,298:INFO:bugbug.model:
Confidence threshold > 0.2 - 5280 classified
Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
Sensitivity is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
                          pre       rec       spe        f1       geo       iba       sup

                 1       0.59      0.76      0.99      0.67      0.87      0.74       122
                 0       0.99      0.99      0.76      0.99      0.87      0.77      5158
__NOT_CLASSIFIED__       0.00      0.00      1.00      0.00      0.00      0.00         0

       avg / total       0.99      0.98      0.77      0.98      0.87      0.77      5280

╒════════════╤═════════════════╤═════════════════╤══════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │   Not classified │
╞════════════╪═════════════════╪═════════════════╪══════════════════╡
│ 1 (Actual) │              93 │              29 │                0 │
├────────────┼─────────────────┼─────────────────┼──────────────────┤
│ 0 (Actual) │              64 │            5094 │                0 │
╘════════════╧═════════════════╧═════════════════╧══════════════════╛

2023-12-21 22:51:58,037:INFO:bugbug.model:
Confidence threshold > 0.3 - 5280 classified
Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
Sensitivity is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
                          pre       rec       spe        f1       geo       iba       sup

                 1       0.69      0.71      0.99      0.70      0.84      0.69       122
                 0       0.99      0.99      0.71      0.99      0.84      0.73      5158
__NOT_CLASSIFIED__       0.00      0.00      1.00      0.00      0.00      0.00         0

       avg / total       0.99      0.99      0.72      0.99      0.84      0.73      5280

╒════════════╤═════════════════╤═════════════════╤══════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │   Not classified │
╞════════════╪═════════════════╪═════════════════╪══════════════════╡
│ 1 (Actual) │              87 │              35 │                0 │
├────────────┼─────────────────┼─────────────────┼──────────────────┤
│ 0 (Actual) │              40 │            5118 │                0 │
╘════════════╧═════════════════╧═════════════════╧══════════════════╛

2023-12-21 22:52:00,938:INFO:bugbug.model:
Confidence threshold > 0.4 - 5280 classified
Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
Sensitivity is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
                          pre       rec       spe        f1       geo       iba       sup

                 1       0.76      0.71      0.99      0.73      0.84      0.69       122
                 0       0.99      0.99      0.71      0.99      0.84      0.73      5158
__NOT_CLASSIFIED__       0.00      0.00      1.00      0.00      0.00      0.00         0

       avg / total       0.99      0.99      0.72      0.99      0.84      0.73      5280

╒════════════╤═════════════════╤═════════════════╤══════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │   Not classified │
╞════════════╪═════════════════╪═════════════════╪══════════════════╡
│ 1 (Actual) │              87 │              35 │                0 │
├────────────┼─────────────────┼─────────────────┼──────────────────┤
│ 0 (Actual) │              28 │            5130 │                0 │
╘════════════╧═════════════════╧═════════════════╧══════════════════╛

2023-12-21 22:52:03,662:INFO:bugbug.model:
Confidence threshold > 0.6 - 5263 classified
Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
Sensitivity is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
                          pre       rec       spe        f1       geo       iba       sup

                 1       0.84      0.67      1.00      0.75      0.82      0.65       122
                 0       0.99      0.99      0.71      0.99      0.84      0.73      5158
__NOT_CLASSIFIED__       0.00      0.00      1.00      0.00      0.00      0.00         0

       avg / total       0.99      0.99      0.72      0.99      0.84      0.73      5280

╒════════════╤═════════════════╤═════════════════╤══════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │   Not classified │
╞════════════╪═════════════════╪═════════════════╪══════════════════╡
│ 1 (Actual) │              82 │              35 │                5 │
├────────────┼─────────────────┼─────────────────┼──────────────────┤
│ 0 (Actual) │              16 │            5130 │               12 │
╘════════════╧═════════════════╧═════════════════╧══════════════════╛

2023-12-21 22:52:06,392:INFO:bugbug.model:
Confidence threshold > 0.7 - 5241 classified
Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
Sensitivity is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
                          pre       rec       spe        f1       geo       iba       sup

                 1       0.86      0.62      1.00      0.72      0.79      0.60       122
                 0       0.99      0.99      0.71      0.99      0.84      0.73      5158
__NOT_CLASSIFIED__       0.00      0.00      0.99      0.00      0.00      0.00         0

       avg / total       0.99      0.98      0.72      0.99      0.84      0.72      5280

╒════════════╤═════════════════╤═════════════════╤══════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │   Not classified │
╞════════════╪═════════════════╪═════════════════╪══════════════════╡
│ 1 (Actual) │              76 │              35 │               11 │
├────────────┼─────────────────┼─────────────────┼──────────────────┤
│ 0 (Actual) │              12 │            5118 │               28 │
╘════════════╧═════════════════╧═════════════════╧══════════════════╛

2023-12-21 22:52:09,060:INFO:bugbug.model:
Confidence threshold > 0.8 - 5204 classified
Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
Sensitivity is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
                          pre       rec       spe        f1       geo       iba       sup

                 1       0.85      0.57      1.00      0.68      0.75      0.54       122
                 0       0.99      0.99      0.76      0.99      0.87      0.77      5158
__NOT_CLASSIFIED__       0.00      0.00      0.99      0.00      0.00      0.00         0

       avg / total       0.99      0.98      0.77      0.98      0.86      0.76      5280

╒════════════╤═════════════════╤═════════════════╤══════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │   Not classified │
╞════════════╪═════════════════╪═════════════════╪══════════════════╡
│ 1 (Actual) │              69 │              29 │               24 │
├────────────┼─────────────────┼─────────────────┼──────────────────┤
│ 0 (Actual) │              12 │            5094 │               52 │
╘════════════╧═════════════════╧═════════════════╧══════════════════╛

2023-12-21 22:52:11,764:INFO:bugbug.model:
Confidence threshold > 0.9 - 5144 classified
Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
Sensitivity is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
                          pre       rec       spe        f1       geo       iba       sup

                 1       0.85      0.52      1.00      0.65      0.72      0.50       122
                 0       1.00      0.98      0.83      0.99      0.90      0.82      5158
__NOT_CLASSIFIED__       0.00      0.00      0.97      0.00      0.00      0.00         0

       avg / total       0.99      0.97      0.83      0.98      0.90      0.81      5280

╒════════════╤═════════════════╤═════════════════╤══════════════════╕
│            │   1 (Predicted) │   0 (Predicted) │   Not classified │
╞════════════╪═════════════════╪═════════════════╪══════════════════╡
│ 1 (Actual) │              64 │              21 │               37 │
├────────────┼─────────────────┼─────────────────┼──────────────────┤
│ 0 (Actual) │              11 │            5048 │               99 │
╘════════════╧═════════════════╧═════════════════╧══════════════════╛

2023-12-21 22:52:12,794:INFO:__main__:Training done
2023-12-21 22:52:14,263:INFO:__main__:Model compressed
```

</details>

Reply:
The results seem better without it: https://github.com/mozilla/bugbug/pull/3775/checks?check_run_id=19873454775

Reply:
Yes, I realized that too. I left it out.